File: code/waterJugProblem.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

public class waterJugProblem{
    public static Boolean solution(int x,int y,int z){
        int [] directions = {x,-x,y,-y};
        if(x+y<z){
            return false;
        }
        Queue<Integer> q = new ArrayDeque<>();
        Set<Integer> visited = new HashSet<>();
        visited.add(0);
        q.offer(0);
        while(!q.isEmpty()){
            int curr = q.poll();
            if(curr == z){
                return true;
            }
            for(int direction: directions ){
                int total = curr + direction;
                if(total<0 || total>=x+y){
                    continue;
                }
                if(total == z){
                    return true;
                }
                if(!visited.contains(total)){
                    visited.add(total);
                    q.offer(total);
                }
            }

        }
        return false;
    }
    public static void main(String[] args) {
        int x=3,y=5,z=4;
        System.out.println(solution(x,y,z)); 
    }
}

File: code/water_jug_problem.cpp
#include<bits/stdc++.h>
#include<algorithm>
using namespace std;
#define ll long long int

bool solution(int x,int y,int z){
    if(x+y<z){
        return false;
    }
    vector<int> direction = {x,-x,y,-y};
    set<int> s;
    queue<int> q;
    s.insert(0);
    q.push(0);
    while(!q.empty()){
        int curr = q.front();
        if(curr == z){
            return true;
        }
        q.pop();
        for(auto directions: direction){
            int total = curr + directions;
            if(total == z){
                return true;
            }
            if(total<0 || total>x+y){
                continue;
            }
            if(s.find(total) == s.end()){
                s.insert(total);
                q.push(total);
            }
        }
    }

    return false;
}

int main()
{
    int x=3;
    int y=5;
    int z=4;
    cout<<solution(x,y,z);
}